pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -347,7 +347,7 @@ func (c regexCache) lookupOrCompileRegex(pattern string) (*regexp.Regexp, error)
 		entry.referenceCount += 1
 		return entry.regex, nil
 	}
-	regex, err := re.CompileRegex(pattern)
+	regex, err := regexp.Compile(pattern)
```

Comment:
I think we can update the godoc on re.CompileRegex to say it's deprecated and not to use it.

Reply:
We can then remove this code in the v1.14 release cycle after we've deprecated it properly here.

Reply:
On second thought, maybe we don't want to entirely remove the LRU.

The problem with caching regexes as we've learned is if we don't clean them up. In your fixes, you've attached a refcount to the regex in the DNS proxy. But there are other areas too.

The other areas for regex compilation is policy YAML validation, aka `pkg/fqdn/matchpattern.Validate()` and [during each](https://github.com/cilium/cilium/blob/b4794d690b5690d70c074bffd1db7593e3938e65/daemon/cmd/fqdn.go#L581) DNS request in the DNS proxy. 

I see value in caching the regex in the above cases. In the former case, we could mitigate the issue of keeping the regexes around by hooking into the policy delete event and removing the names (which map to regexes) in the policy from the cache. In the latter case, we can use an LRU. I feel an LRU is suitable in this case because the access pattern will be based on the DNS requests that the workloads are performing.

Should these cases share the same LRU? Well, that was kinda the idea behind the initial LRU implementation to begin with. I think as long as we manage the lifecycle and ensure we delete "stale" entries, then we will probably be fine regarding all the memory issues we've had in the past (and that you've helped fix :slightly_smiling_face: ).